File: src/lessons/recursion/hanoi/IterativeHanoiEntity.java
package lessons.recursion.hanoi;

import plm.universe.hanoi.HanoiEntity;

public class IterativeHanoiEntity extends HanoiEntity {
	
	public void run() {
		hanoi((Integer)getParam(0),(Boolean) getParam(1));
	}

	/* BEGIN TEMPLATE */
	void hanoi(int initialPos, boolean increasing) {
		/* BEGIN SOLUTION */
		int small = initialPos;
		int count = 0;
		int pos1=0,pos2=0;
		do {
			if (count%2 == 0) {
				int next = (small+3+(increasing?1:-1)) % 3;
				//getGame().getLogger().log("move("+small+","+next+")");
				move(small, next);
				small=next;
			}
			
			switch (small) {
			case 0: pos1=1; pos2=2; break;
			case 1: pos1=0; pos2=2; break;
			case 2: pos1=0; pos2=1; break;
			}
			if (count%2 == 1) {
				if (getSlotRadius(pos1) > getSlotRadius(pos2))
					move(pos2,pos1);
				else
					move(pos1,pos2);
			}
			
			count++;
		} while (getSlotSize(pos1) != 0 || getSlotSize(pos2) != 0);
		/* END SOLUTION */
	}
	/* END TEMPLATE */

}

File: resources/exercises/recursion/hanoi/IterativeHanoiEntity.py
# BEGIN TEMPLATE
def hanoi(initialPos, increasing):
	
	# Your code here
	# BEGIN SOLUTION
	count=0
	small=initialPos
	p1=initialPos
	p2=initialPos
	while getSlotSize(p1)!=0 or getSlotSize(p2)!=0:
		if (count%2 == 0):
			if increasing:
				next = (small+1)%3
			else:
				next = (small-1+3)%3
			move(small,next)
			small = next
		if small == 0:
			p1=1
			p2=2
		elif small == 1:
			p1=0
			p2=2
		else:
			p1=0
			p2=1
		if (count%2 == 1):
			if (getSlotRadius(p1)> getSlotRadius(p2)):
				move(p2,p1)
			else:
				move(p1,p2)
		count += 1			
	# END SOLUTION
# END TEMPLATE
hanoi(getParam(0), getParam(1))